File: lib/src/main/java/org/simplepresenter/LcePresenterFragment.java
package org.simplepresenter;

import android.support.annotation.NonNull;

import org.simplepresenter.commands.DataCommand;
import org.simplepresenter.commands.ErrorViewCommand;
import org.simplepresenter.commands.ProgressViewCommand;

public abstract class LcePresenterFragment<D, P extends Presenter> extends PresenterFragment<P> {

    @Override
    public void dispatchCommand(@NonNull ViewCommand viewCommand) {
        if (viewCommand instanceof ProgressViewCommand) {
            onProgress();
        } else if (viewCommand instanceof ErrorViewCommand) {
            onError(((ErrorViewCommand) viewCommand).getThrowable());
        } else if (viewCommand instanceof DataCommand<?>) {
            //noinspection unchecked
            onData((D) ((DataCommand) viewCommand).getData());
        }
    }

    public void onProgress() {
    }

    public void onError(Throwable throwable) {
    }

    public void onData(D data) {
    }
}

File: lib/src/main/java/org/simplepresenter/LcePresenterFragment.kt
package org.simplepresenter

import org.simplepresenter.commands.DataCommand
import org.simplepresenter.commands.ErrorViewCommand
import org.simplepresenter.commands.ProgressViewCommand

@Suppress("unused")
abstract class LcePresenterFragment<in D, out P : Presenter> : PresenterFragment<P>() {

    override final fun dispatch(viewCommand: ViewCommand) {
        if (viewCommand is ProgressViewCommand) {
            onProgress()
        } else if (viewCommand is ErrorViewCommand) {
            onError(viewCommand.throwable)
        } else if (viewCommand is DataCommand<*>) {
            @Suppress("UNCHECKED_CAST")
            onData(viewCommand.data as D)
        } else {
            onElse(viewCommand)
        }
    }

    fun onProgress() {
    }

    fun onError(throwable: Throwable) {
    }

    fun onData(data: D) {
    }

    fun onElse(viewCommand: ViewCommand) {
    }
}